### PR TITLE
Wrap observable source assets in AssetsDefinition

### PR DESCRIPTION
**Pull request by schrockn**
## Summary & Motivation

This makes an observable source asset wrappable in an AssetsDefinition. This is an imperfect intermediate state, as as spurious asset materialization is produced as a result of execution (verified in test case but fixed in https://github.com/dagster-io/dagster/pull/16621) but this verifies that the core execution machinery works.

Makes the following code possible:

```python
@observable_source_asset
def an_observable_source_asset() -> DataVersion:
    return DataVersion("foo")

defs = Definitions(assets=[
    create_external_asset_from_source_asset(an_observable_source_asset)
])
```

This `@observable_source_asset` now plugs into the rest of our infrastructure:

1) You can execute it from `dagster-webserver` using the "Materialize" button, which will be renamed to "Execute".
2) You can schedule this from schedules and sensors.

AMP support has to wait because auto observation is hard-coded in the AMP logic.

## How I Tested These Changes

BK

### CODE REVIEW

File: python_modules/dagster/dagster/_core/definitions/observable_asset.py [hunk 1]
```diff
@@ -79,8 +86,17 @@ def create_unexecutable_observable_assets_def_from_source_asset(source_asset: So
...
+
     @asset(**kwargs)
-    def shim_asset() -> None:
-        raise NotImplementedError(f"Asset {source_asset.key} is not executable")
+    def shim_asset(context: OpExecutionContext) -> None:
+        if not source_asset.observe_fn:
+            raise NotImplementedError(f"Asset {source_asset.key} is not executable")
+
+        op_function = wrap_source_asset_observe_fn_in_op_compute_fn(source_asset)
+        return op_function.decorated_fn(context)
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/definitions/observable_asset.py:100]: what happens output wise for observable source assets? Does the default io manager store whatever dataversion object gets returned or do we have special casing somewhere?

**Review by alangenfeld — COMMENTED** [review]
see inline - I think we might want more explicit testing against the current execution pattern for existing `@observable_source_asset` jobs compared to new approach

**Review by alangenfeld — CHANGES_REQUESTED** [review]
i think ^ is worth taking a closer look at

File: python_modules/dagster/dagster/_core/definitions/observable_asset.py (hunk 1 above)
Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/observable_asset.py]: Confirmed that this does not result in the emission of an Output.

1) `DataVersion` gets translated to an `AssetObservation` in `wrap_source_asset_observe_fn_in_op_compute_fn`
2) `wrap_source_asset_observe_fn_in_op_compute_fn` also creates a compute fn guaranteed to return None.

2 is not enforced by the type system, but I have added an invariant with a helpful message to describe the state of things

Also added a test to confirm this behavior.

**Review by alangenfeld — APPROVED** [review]

**Pull request merged**